pki/certs.go

```diff
@@ -0,0 +1,111 @@
+package pki
+
+import (
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/hex"
+	"errors"
+	"math"
+	"math/big"
+
+	"github.com/jmhodges/clock"
+)
+
+const (
+	// Domain suffix for the subject common name of certificates generated for submission
+	// to logs. The prefix will be generated randomly from the certificate serial number.
+	testCertDomain = ".woodpecker.testing.letsencrypt.org"
+)
+
+// RandSerial generates a random *bigInt to use as a certificate serial or
+// returns an error.
+func RandSerial() (*big.Int, error) {
+	serial, err := rand.Int(rand.Reader, big.NewInt(math.MaxInt64))
+	if err != nil {
+		return nil, err
+	}
+	return serial, nil
+}
+
+// RandKey generates a random ECDSA private key or returns an error.
+func RandKey() (*ecdsa.PrivateKey, error) {
+	key, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	if err != nil {
+		return nil, err
+	}
+	return key, nil
+}
+
+// IssueCertificate uses the monitor's certIssuer and certIssuerKey to generate
+// a leaf-certificate that can be submitted to a log. The certificate's subject
+// common name will be the monitor's subjCNPrefix and the first three bytes of
+// the random serial number encoded in hex.
+// TODO(@cpu) Write this comment
+
+func IssueCertificate(
+	subjectKey crypto.PublicKey,
+	issuerKey *ecdsa.PrivateKey,
+	issuerCert, template *x509.Certificate) (*x509.Certificate, error) {
+	if subjectKey == nil {
+		return nil, errors.New("cannot IssueCertificate with nil subjectKey")
+	}
+	if issuerKey == nil {
+		return nil, errors.New("cannot IssueCertificate with nil issuerKey")
+	}
+	if issuerCert == nil {
+		return nil, errors.New("cannot IssueCertificate with nil issuerCert")
+	}
+	if template == nil {
+		return nil, errors.New("cannot IssueCertificate with nil template")
+	}
+
+	certDER, err := x509.CreateCertificate(rand.Reader, template, issuerCert, subjectKey, issuerKey)
+	if err != nil {
+		return nil, err
+	}
+
+	cert, err := x509.ParseCertificate(certDER)
+	if err != nil {
+		return nil, err
+	}
+	return cert, nil
+}
+
+func IssueTestCertificate(
```

Comment:
Add a comment here? In particular I think it's worth calling out that this function creates certificates that will be submitted to public logs, so even though the certificates are not signed by a trusted root, we try to generally pass cablint to avoid making people who run monitors add special cases.